fabric-registry-sync-v0/src/client/java/net/fabricmc/fabric/impl/client/registry/sync/RegistryRemovalChecker.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.client.registry.sync;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.mojang.datafixers.util.Pair;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.minecraft.block.Block;
+import net.minecraft.registry.Registries;
+import net.minecraft.registry.Registry;
+import net.minecraft.registry.RegistryKey;
+import net.minecraft.registry.RegistryKeys;
+import net.minecraft.state.property.Property;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.JsonHelper;
+import net.minecraft.util.Util;
+
+import net.fabricmc.fabric.api.event.registry.RegistryAttribute;
+import net.fabricmc.fabric.api.event.registry.RegistryAttributeHolder;
+
+public final class RegistryRemovalChecker {
+	private static final int VERSION = 1;
+	public static final Logger LOGGER = LoggerFactory.getLogger(RegistryRemovalChecker.class);
+	private static final Gson GSON = new Gson();
+	public static final String FILE_NAME = "fabric_registry_removal_check.json";
+	private static final boolean DISABLED = Boolean.getBoolean("fabric.registry.debug.disableRemovalCheck");
+	private final Set<String> missingNamespaces;
+	private final Set<RegistryKey<?>> missingKeys;
+	private final Set<Pair<Block, String>> missingBlockStates = new HashSet<>();
+
+	@SuppressWarnings("unchecked, rawtypes")
+	public RegistryRemovalChecker(JsonObject root) {
+		JsonObject json = root.getAsJsonObject("entries");
```

Comment:
This never checks the version, it should handle the case of the version being newer than supported. I imagine telling the user this.